### PR TITLE
H7 low power fixes

### PR DESCRIPTION
**Pull request by iabdalkader**
This PR brings some fixes to the low power stop/standby modes on the H7:

* Fix STOP mode voltage scaling on REV V devices.
* Re-enable disabled OSCs and PLLs after exiting stop mode.
* Disable RTC register write protection before changing flags.
* Set H7 RTC wakeup flags.
* Define missing RCC_SR_SFTRSTF flag for H747.
* Make sure Systick is disabled on stop mode entry for H7 REV Y devices.

Split into smaller commits for cherry-picking. 
Note I tested standby wake up on RTC and stop mode wake up on RTC and extint on both REV Y and V.

### CODE REVIEW

**Comment by iabdalkader** [comment]
Alternatively we can check for individual OSCs and PLLs instead of H7 for other micros that have these defined:

```c
diff --git a/ports/stm32/powerctrl.c b/ports/stm32/powerctrl.c
index 076215ba9..62cf56444 100644
--- a/ports/stm32/powerctrl.c
+++ b/ports/stm32/powerctrl.c
@@ -516,10 +516,11 @@ void powerctrl_enter_stop_mode(void) {
     HAL_PWREx_EnableFlashPowerDown();
     #endif
 
-    #if defined(STM32H7)
     // Save RCC CR to re-enable OSCs and PLLs after wake up from low power mode.
     uint32_t rcc_cr = RCC->CR;
+    (void) rcc_cr;
 
+    #if defined(STM32H7)
     // Save the current voltage scaling level to restore after exiting low power mode.
     uint32_t vscaling = POWERCTRL_GET_VOLTAGE_SCALING();
 
@@ -609,35 +610,43 @@ void powerctrl_enter_stop_mode(void) {
     }
     #endif
 
-    #if defined(STM32H7)
+    #if defined(RCC_CR_HSION)
     // Enable HSI
     if (rcc_cr & RCC_CR_HSION) {
         RCC->CR |= RCC_CR_HSION;
         while (!(RCC->CR & RCC_CR_HSIRDY)) {
         }
     }
+    #endif
 
+    #if defined(RCC_CR_CSION)
     // Enable CSI
     if (rcc_cr & RCC_CR_CSION) {
         RCC->CR |= RCC_CR_CSION;
         while (!(RCC->CR & RCC_CR_CSIRDY)) {
         }
     }
+    #endif
 
+    #if defined(RCC_CR_HSI48ON)
     // Enable HSI48
     if (rcc_cr & RCC_CR_HSI48ON) {
         RCC->CR |= RCC_CR_HSI48ON;
         while (!(RCC->CR & RCC_CR_HSI48RDY)) {
         }
     }
+    #endif
 
+    #if defined(RCC_CR_PLL2ON)
     // Enable PLL2
     if (rcc_cr & RCC_CR_PLL2ON) {
         RCC->CR |= RCC_CR_PLL2ON;
         while (!(RCC->CR & RCC_CR_PLL2RDY)) {
         }
     }
+    #endif
 
+    #if defined(RCC_CR_PLL3ON)
     // Enable PLL3
     if (rcc_cr & RCC_CR_PLL3ON) {
         RCC->CR |= RCC_CR_PLL3ON;
```

**Comment by iabdalkader** [comment]
Pushed one more commit to fix this:

```
Breakpoint 1, powerctrl_enter_stop_mode () at powerctrl.c:500
500	void powerctrl_enter_stop_mode(void) {
(gdb) n
503	    uint32_t irq_state = disable_irq();
(gdb) 
521	    HAL_PWREx_EnableFlashPowerDown();
(gdb) enable 2
(gdb) n
526	    uint32_t rcc_cr = RCC->CR;
(gdb) 
529	    uint32_t vscaling = POWERCTRL_GET_VOLTAGE_SCALING();
(gdb) n
532	    if (vscaling == PWR_REGULATOR_VOLTAGE_SCALE0) {
(gdb) 
543	    HAL_PWR_EnterSTOPMode(PWR_LOWPOWERREGULATOR_ON, PWR_STOPENTRY_WFI);
(gdb) c
Continuing.

Breakpoint 2, SysTick_Handler () at systick.c:39
39	void SysTick_Handler(void) {
```

File: ports/stm32/powerctrl.c [hunk 1]
```diff
@@ -492,6 +502,11 @@ void powerctrl_enter_stop_mode(void) {
     // executed until after the clocks are reconfigured
     uint32_t irq_state = disable_irq();
 
+    // Disable SysTick Interrupt
+    // Note: This seems to be required at least on the H7 REV Y,
+    // otherwise the MCU will leave stop mode immediately on entry.
+    SysTick->CTRL &= ~SysTick_CTRL_TICKINT_Msk;
```
Comment (dpgeorge) [inline ports/stm32/powerctrl.c:508]: This is strange.... it means that the SysTick IRQ can now wake from STOP mode.  Is there anything in the datasheet about this?

File: ports/stm32/powerctrl.c [hunk 2]
```diff
@@ -622,6 +697,10 @@ void powerctrl_enter_standby_mode(void) {
     // save RTC interrupts
     uint32_t save_irq_bits = RTC->CR & CR_BITS;
 
+    // disable register write protection
+    RTC->WPR = 0xca;
+    RTC->WPR = 0x53;
```
Comment (dpgeorge) [inline ports/stm32/powerctrl.c:702]: Why is this needed?

**Comment by dpgeorge** [comment]
> Alternatively we can check for individual OSCs and PLLs instead of H7 for other micros that have these defined

I think the first way you did it is good for now.  Later it can be changed to have each individual part selected based on availability of that OSC/PLL, if needed.

File: ports/stm32/powerctrl.c (hunk 1 above)
Comment (iabdalkader) [reply ports/stm32/powerctrl.c:508]: No not that I could find, I searched for a while and checked the latest errata, I think it may be an undocumented silicon bug because this does Not happen on H743 REV-V. SysTick IRQ gets disabled with the other IRQs as you can see from the gdb log, but then triggers on entry to STOP mode. Disabling it like that shouldn't have any side effect on other MCUs though.

File: ports/stm32/powerctrl.c (hunk 2 above)
Comment (iabdalkader) [reply ports/stm32/powerctrl.c:702]: Because the RTC registers are left write-protected in `pyb_rtc_wakeup`, and these writes in `powerctrl_enter_standby_mode`, useful or not, have no effect:
```
Breakpoint 1, powerctrl_enter_standby_mode () at powerctrl.c:674
674 void powerctrl_enter_standby_mode(void) {
705     RTC->CR &= ~CR_BITS;
(gdb) 
708     RTC->ISR &= ~ISR_BITS;
(gdb) print /x RTC->CR
$18 = 0x44402
(gdb) set RTC->WPR = 0xca
(gdb) set RTC->WPR = 0x53
(gdb) set RTC->CR &= ~CR_BITS
(gdb) print /x RTC->CR
$19 = 0x40402
```

File: ports/stm32/powerctrl.c (hunk 1 above)
Comment (iabdalkader) [reply ports/stm32/powerctrl.c:508]: Actually after thinking about it some more, it may not be disabled at all because when single stepping with the debugger systick counter is suspended as far as I know.

Comment (dpgeorge) [reply ports/stm32/powerctrl.c:508]: Ok.  But I'd say this should be guarded by `defined(STM32H7)` because this cause is not properly understood.

File: ports/stm32/powerctrl.c (hunk 2 above)
Comment (dpgeorge) [reply ports/stm32/powerctrl.c:702]: Ah, I see!  Yes it's needed.

File: ports/stm32/powerctrl.c (hunk 1 above)
Comment (iabdalkader) [reply ports/stm32/powerctrl.c:509]: Done.

**Comment by dpgeorge** [comment]
Thanks for these, they are good changes.

Merged in e9e619fa24cca3aa6788931d6fee561436576f83 through 7dc2f4ed384f45eb943e0b7dc39a539af7058b80
